File: sdks/cpp/tests/sak_client_test.cpp
#include <catch2/catch_test_macros.hpp>
#include <httplib.h>
#include <thread>

#include "swissarmynoife/sak_client.hpp"

TEST_CASE("SakClient strips slash", "[http]") {
  swissarmynoife::SakClient c("http://127.0.0.1:8787/");
  REQUIRE(c.base_url() == "http://127.0.0.1:8787");
}

TEST_CASE("SakClient health and lists", "[http]") {
  httplib::Server svr;
  svr.Get("/health", [](const httplib::Request&, httplib::Response& res) {
    res.set_content(R"({"ok":true})", "application/json");
  });
  svr.Get("/v1/sak/modules", [](const httplib::Request&, httplib::Response& res) {
    res.set_content(R"({"modules":[]})", "application/json");
  });
  svr.Get("/v1/sak/capacity", [](const httplib::Request&, httplib::Response& res) {
    res.set_content(R"({"snapshot":{"total_ram_mb":1}})", "application/json");
  });
  svr.Get("/v1/sak/compute/work", [](const httplib::Request&, httplib::Response& res) {
    res.set_content(R"({"work":[]})", "application/json");
  });
  svr.Get("/v1/sak/compute/nodes", [](const httplib::Request&, httplib::Response& res) {
    res.set_content(R"({"nodes":[]})", "application/json");
  });
  auto port = svr.bind_to_any_port("127.0.0.1");
  std::thread t([&] { svr.listen_after_bind(); });
  swissarmynoife::SakClient c("http://127.0.0.1:" + std::to_string(port));
  REQUIRE(c.health()["ok"] == true);
  REQUIRE(c.list_modules().contains("modules"));
  REQUIRE(c.capacity().contains("snapshot"));
  REQUIRE(c.list_work().contains("work"));
  REQUIRE(c.list_nodes().contains("nodes"));
  svr.stop();
  t.join();
}

TEST_CASE("SakClient enqueue and claim", "[http]") {
  std::string last;
  httplib::Server svr;
  svr.Post("/v1/sak/compute/work", [&](const httplib::Request& req, httplib::Response& res) {
    last = req.body;
    res.set_content(R"({"action":"ok","work":{"id":"w1"}})", "application/json");
  });
  auto port = svr.bind_to_any_port("127.0.0.1");
  std::thread t([&] { svr.listen_after_bind(); });
  swissarmynoife::SakClient c("http://127.0.0.1:" + std::to_string(port));
  auto out = c.enqueue_work("echo", {{"n", 1}});
  REQUIRE(out["action"] == "ok");
  REQUIRE(last.find("\"action\":\"enqueue\"") != std::string::npos);
  c.requeue_work("w1");
  REQUIRE(last.find("requeue") != std::string::npos);
  c.claim_work("n1");
  REQUIRE(last.find("claim") != std::string::npos);
  c.complete_work("w1", "n1");
  REQUIRE(last.find("complete") != std::string::npos);
  c.get_work("w1");
  REQUIRE(last.find("\"action\":\"get\"") != std::string::npos);
  svr.stop();
  t.join();
}

File: sdks/php/tests/SakClientTest.php
<?php

declare(strict_types=1);

namespace SwissArmyNoife\Sdk\Tests;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use GuzzleHttp\Psr7\Response;
use PHPUnit\Framework\TestCase;
use SwissArmyNoife\Sdk\SakClient;

final class SakClientTest extends TestCase
{
    public function testBaseUrlStripsSlash(): void
    {
        $c = new SakClient('http://127.0.0.1:8787/');
        $this->assertSame('http://127.0.0.1:8787', $c->baseUrl());
    }

    public function testHealth(): void
    {
        $history = [];
        $c = $this->clientWithResponses(
            [new Response(200, [], '{"ok":true}')],
            $history,
        );
        $this->assertTrue($c->health()['ok']);
        $this->assertSame('/health', $history[0]['request']->getUri()->getPath());
    }

    /** @dataProvider listHelpers */
    public function testListHelpers(string $method, string $path, string $body): void
    {
        $history = [];
        $c = $this->clientWithResponses([new Response(200, [], $body)], $history);
        $out = $c->{$method}();
        $this->assertIsArray($out);
        $this->assertSame($path, $history[0]['request']->getUri()->getPath());
    }

    /** @return array<string, array{0:string,1:string,2:string}> */
    public static function listHelpers(): array
    {
        return [
            'modules' => ['listModules', '/v1/sak/modules', '{"modules":[]}'],
            'work' => ['listWork', '/v1/sak/compute/work', '{"work":[]}'],
            'nodes' => ['listNodes', '/v1/sak/compute/nodes', '{"nodes":[]}'],
            'capacity' => ['capacity', '/v1/sak/capacity', '{"snapshot":{"total_ram_mb":1}}'],
        ];
    }

    public function testEnqueueWork(): void
    {
        $history = [];
        $c = $this->clientWithResponses(
            [new Response(200, [], '{"action":"enqueue","work":{"status":"queued"}}')],
            $history,
        );
        $out = $c->enqueueWork('echo', ['n' => 1]);
        $this->assertSame('enqueue', $out['action']);
        $payload = json_decode((string) $history[0]['request']->getBody(), true);
        $this->assertSame('enqueue', $payload['action']);
        $this->assertSame('echo', $payload['kind']);
    }

    public function testRequeueClaimCompleteGet(): void
    {
        $history = [];
        $responses = array_fill(
            0,
            4,
            new Response(200, [], '{"action":"ok","work":{"id":"w1"}}'),
        );
        $c = $this->clientWithResponses($responses, $history);
        $c->requeueWork('w1');
        $c->claimWork('n1');
        $c->completeWork('w1', 'n1');
        $c->getWork('w1');
        $actions = array_map(
            static fn ($h) => json_decode((string) $h['request']->getBody(), true)['action'],
            $history,
        );
        $this->assertSame(['requeue', 'claim', 'complete', 'get'], $actions);
    }

    /**
     * @param list<Response> $responses
     * @param list<array<string, mixed>> $history
     */
    private function clientWithResponses(array $responses, array &$history): SakClient
    {
        $mock = new MockHandler($responses);
        $stack = HandlerStack::create($mock);
        $stack->push(Middleware::history($history));

        return new SakClient('http://example.test', new Client(['handler' => $stack]));
    }
}
